### PR TITLE
[Merged by Bors] - Publish subscriptions to all beacon nodes

### LINKED ISSUE #3516: VC: Publish to multiple BNs
I have my lighthouse VC connected to 2 BNs (primary and fallback).
I propose to add a flag that makes the VC publish messages via _all_ connected BNs concurrently.
That would sidestep bad fall-back experiences (see [comment in 3002](https://github.com/sigp/lighthouse/issues/3002#issuecomment-1229185227)) with missed attestions/proposals/sync committees.
It would also increase the chances of messages being successfully published despite spotty internet connections with occasional packet loss on the beacon nodes.

Implementing this would make #3001 and #3002 not required anymore for solid fall-back.

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

Resolves #3516 

## Proposed Changes

Adds a beacon fallback function for running a beacon node http query on all available fallbacks instead of returning on a first successful result. Uses the new `run_on_all` method for attestation and sync committee subscriptions. 

## Additional Info

Please provide any additional information. For example, future considerations
or information useful for reviewers.

### CODE REVIEW

**Comment by torfbolt** [comment]
Also resolves #3422

File: validator_client/src/cli.rs [hunk 1]
```diff
@@ -26,6 +26,16 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
                 )
                 .takes_value(true),
         )
+        .arg(
+            Arg::with_name("disable-publish-subsriptions-all")
```
Comment (michaelsproul) [inline validator_client/src/cli.rs]:
```suggestion
            Arg::with_name("disable-publish-subscriptions-all")
```

File: validator_client/src/config.rs [hunk 2]
```diff
@@ -181,6 +184,8 @@ impl Config {
         }
 
         config.allow_unsynced_beacon_node = cli_args.is_present("allow-unsynced");
+        config.disable_publish_subscriptions_all =
+            cli_args.is_present("disable-publish-subsriptions-all");
```
Comment (michaelsproul) [inline validator_client/src/config.rs]:
```suggestion
            cli_args.is_present("disable-publish-subscriptions-all");
```

**Review by michaelsproul — COMMENTED** [review]
Just a quick passing review

Would be good to have tests for the CLI flag in `lighthouse/tests/validator_client.rs`

File: validator_client/src/beacon_node_fallback.rs [hunk 3]
```diff
@@ -105,11 +105,13 @@ impl<E> Error<E> {
...
 
 /// The list of errors encountered whilst attempting to perform a query.
-pub struct AllErrored<E>(pub Vec<(String, Error<E>)>);
+pub struct Errors<E>(pub Vec<(String, Error<E>)>);
 
-impl<E: Debug> fmt::Display for AllErrored<E> {
+impl<E: Debug> fmt::Display for Errors<E> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "All endpoints failed")?;
+        if self.0.len() > 1 {
```
Comment (realbigsean) [inline validator_client/src/beacon_node_fallback.rs]: We could write out the number failed here right?

File: validator_client/src/beacon_node_fallback.rs [hunk 4]
```diff
@@ -486,6 +488,123 @@ impl<T: SlotClock, E: EthSpec> BeaconNodeFallback<T, E> {
         }
 
         // There were no candidates already ready and we were unable to make any of them ready.
-        Err(AllErrored(errors))
+        Err(Errors(errors))
+    }
+
+    /// Run `func` against all candidates in `self`, collecting the result of func against each
```
Comment (realbigsean) [inline validator_client/src/beacon_node_fallback.rs]:
```suggestion
    /// Run `func` against all candidates in `self`, collecting the result of `func` against each
```

File: validator_client/src/cli.rs [hunk 5]
```diff
@@ -26,6 +26,16 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
                 )
                 .takes_value(true),
         )
+        .arg(
+            Arg::with_name("disable-publish-subscriptions-all")
+                .long("disable-publish-subsriptions-all")
```
Comment (realbigsean) [inline validator_client/src/cli.rs]:
```suggestion
                .long("disable-publish-subscriptions-all")
```

File: validator_client/src/duties_service.rs [hunk 6]
```diff
@@ -570,26 +571,44 @@ async fn poll_beacon_attesters<T: SlotClock + 'static, E: EthSpec>(
...
-                        &[metrics::SUBSCRIPTIONS_HTTP_POST],
-                    );
-                    beacon_node
-                        .post_validator_beacon_committee_subscriptions(subscriptions_ref)
-                        .await
-                },
-            )
-            .await
-        {
+        if let Err(e) = if duties_service.disable_run_on_all {
```
Comment (realbigsean) [inline validator_client/src/duties_service.rs]: What're your thoughts on adding a really simple function to `BeaconNodeFallback` so we don't have to have these `if` `else` blocks?

```
    pub async fn run<'a, F, Err, R>(
        &'a self,
        require_synced: RequireSynced,
        offline_on_failure: OfflineOnFailure,
        func: F,
        disable_run_on_all: bool,
    ) -> Result<(), Errors<Err>>
        where
            F: Fn(&'a BeaconNodeHttpClient) -> R,
            R: Future<Output = Result<(), Err>>,
    {
        if disable_run_on_all {
            self.first_success(require_synced, offline_on_failure, func).await?;
            Ok(())
        } else {
            self.run_on_all(require_synced, offline_on_failure, func).await
        }
    }
```

We could also add the `disable_run_on_all` bool to `BeaconNodeFallback` so we don't have to pass it in as an argument

**Review by realbigsean — CHANGES_REQUESTED** [review]

File: validator_client/src/duties_service.rs (hunk 6 above)
Comment (pawanjay176) [reply validator_client/src/duties_service.rs]: I like this idea. Will make the changes.

File: validator_client/src/beacon_node_fallback.rs (hunk 3 above)
Comment (pawanjay176) [reply validator_client/src/beacon_node_fallback.rs]: Done!

**Review by realbigsean — APPROVED** [review]
Nice!

**Pull request closed without merging**